### PR TITLE
Cache GET requests to OpenID providers.

### LINKED ISSUE #6: Cache OIDC provider metadata
Can cache the bits of the discovery document we use in Redis. Probably makes sense to just cache with a 3600s expiry, rather than doing complicated ETag and Last-Modified handling? The Google IdP sets timeout of an hour explicitly, though Microsoft doesn't send anything like it.

### PR DESCRIPTION
**Pull request by stephank**
Fixes #6 and #7.

Note that these are still babysteps in Rust for me. Beside that, some things I'm considering:

 - Maybe we should have `fetch_json_cached` be a store method. Seems like a slightly better place to handle this sort of thing than in oidc.

 - Minimum cache age is now not configurable. Simplest would be using `expire_keys`, but it seems weird to tie it to session lifespan. I'm itching to rename `expire_keys`, actually, and then we can have `session_age`, `min_cache_age`, and `max_cache_size` in the future. But I'm not sure what our view is on breaking changes?

### CODE REVIEW

**Comment by djc** [comment]
This is looking great! Have you done some testing on this?

I don't have time to review in detail right now, but here are some high-level remarks: I think the functions you put in the `store` module should probably be part of `impl Store`. Did you have a reason to put them next to it, rather than in it?

I'd prefer not to do the `max-age` thing, since I saw that not all providers have it. I think just doing 1 hour of caching is safe enough in practice, and still provides significant benefit.

Breaking changes are not an issue at all as far as I'm concerned. Maybe you can come up with a detailed proposal for cache changes before you implement it? I don't like the `_age` naming much, for what it's worth, since `expire_` is a bit more explicit about what's going on. However, making the settings more granular makes sense to me.

**Comment by stephank** [comment]
I tested with a Google apps domain, and verified the behavior of the cache and `max-age` handling using `RUST_LOG=debug`.

No solid reason for putting methods outside `impl Store`, except wanting them to be private.

Why would we ignore `max-age` if it's present? Right now, we enforce a minimum of 1 hour, even if `max-age` is shorter or absent. But if a provider wants to instruct us to cache longer, it makes sense to use the standard header for it. (I guess a maximum wouldn't hurt either.)

As for naming, I guess we can be more explicit. Following `expire_`, how about `expire_session`, `expire_cache_minimum`, `maximum_response_size`? (I don't like `_age` either, but tried to stick to the HTTP terminology. Personally would've gone with `_lifespan`.)

My focus was on the two specific issues, so I'm not sure if there are or will be other uses for caching. I'm not sure what the proper balance is between pretty and pragmatic, because I would tend towards the former if I were to write up a detailed proposal.

But in the pragmatic case, we could just iterate on the changes and feedback here.

**Comment by stephank** [comment]
On second thought, maybe we shouldn't have a maximum cache lifespan. I was concerned about a malicious party flooding memory, but think that'd be still possible even with the option.

For example, if you define min / max as 1 hour / 1 week, that'd mean a single domain could serve a config with a rotating keys URL and create 24 * 7 = 168 cache entries for just keys, each up to a maximum of the `maximum_response_size`. Combine that with an attack using a wildcard domain, and you still have a real problem.

Rather, I think we should just instruct the user to properly configure Redis with `maxmemory` and `maxmemory-policy` with probably the `volatile-lru` strategy. (See: http://redis.io/topics/lru-cache)

A minimum cache lifespan still makes sense to reduce traffic, of course. And a maximum response size is always sensible, and prevents an instant denial of service.

**Comment by stephank** [comment]
In the last two commits, I move the helpers inside `impl Store`, and move the cache-related code to Store, including the code for basic cached GET-requests.

**Comment by stephank** [comment]
The last four commits are some cleanup, and the implementation of the two remaining config options. I've gone with the following:

```
  "store": {
    "redis_url": "redis://127.0.0.1/5",
    "expire_sessions": 900,
    "expire_cache": 3600,
    "max_response_size": 8096
  },
```

But changing the naming is easy enough, so let me know if that needs to be done.

File: src/store.rs [hunk 1]
```diff
@@ -9,39 +10,31 @@ pub struct Store {
...
-        if stored.is_empty() {
-            return Err("session not found".to_string());
-        }
-        Ok(stored)
+        let key = format_session_key(session_id);
+        self.client.hgetall(key.as_str()).map_err(stringify_err)
     }
 }
+
+fn stringify_err(err: RedisError) -> String {
```
Comment (djc) [inline src/store.rs]: Can you try to do this with a `From` or `Into` implementation, instead? Ideally we would then not need to reference the function (later method) everywhere.

File: src/store.rs [hunk 2]
```diff
@@ -29,6 +29,19 @@ impl Store {
         let key = format_session_key(session_id);
         self.client.hgetall(key.as_str()).map_err(stringify_err)
     }
+
+    pub fn store_cache(&self, url: &str, data: &str)
```
Comment (djc) [inline src/store.rs]: I'm not sure this is the right abstraction for this. Maybe we can add a separate `StoreCache` trait that takes a store, and then store in there? I think it's also better not to instantiate a new `Http::Client` for every request. The `StoreCache` should probably have a single method that takes `type` (so we can disambiguate different types of cached data, instead of prefixing them all with `cache:`), `key`, `session` (an `Http::Client`), and the URL to get, and returns the value.

File: src/oidc.rs [hunk 3]
```diff
@@ -145,18 +150,38 @@ pub fn verify(app: &AppConfig, session_id: &str, code: &str)
...
+        // Consider deserializing/serializing for validation, and to remove unnecessary spaces.
         let mut data = String::new();
         let _ = rsp.read_to_string(&mut data).unwrap();
-        let _ = app.store.store_cache(url, &data).unwrap();
+
+        // Cache the response for at least an hour.
+        // TODO: Consider making this configurable.
+        let cache_life = max(3600, max_age as usize);
+        let _ = app.store.store_cache(url, &data, cache_life).unwrap();
+
```
Comment (djc) [inline src/oidc.rs]: Nit: this empty line feels pointless.

File: src/oidc.rs [hunk 4]
```diff
@@ -89,10 +89,8 @@ pub fn verify(app: &AppConfig, session_id: &str, code: &str)
     let email_addr = EmailAddress::new(stored.get("email").unwrap()).unwrap();
     let origin = stored.get("client_id").unwrap();
 
-    // Request the provider's Discovery document to get the
-    // `token_endpoint` and `jwks_uri` values from it. TODO: save these
-    // when requesting the Discovery document in the `oauth_request()`
-    // function, and/or cache them by provider host.
+    // Retrieve the provider's Discovery document to get the
```
Comment (djc) [inline src/oidc.rs]: Please squash these cleanups into the commits where you actually fixed it.

File: src/store.rs [hunk 5]
```diff
@@ -12,46 +12,46 @@ pub struct Store {
...
-        self.client.get(key.as_str()).map_err(stringify_err)
+        let key = Self::format_cache_key(url);
+        self.client.get(key.as_str()).map_err(Self::stringify_err)
     }
-}
 
-fn stringify_err(err: RedisError) -> String {
-    err.description().to_string()
-}
+    fn stringify_err(err: RedisError) -> String {
```
Comment (djc) [inline src/store.rs]: Squash this into the commit that introduces the function, please (only if the `From`/`Into` implementation doesn't work out; that should also be part of the same commit).

File: src/store.rs [hunk 6]
```diff
@@ -12,46 +12,46 @@ pub struct Store {
...
-fn format_session_key(session_id: &str) -> String {
-    format!("session:{}", session_id)
-}
+    fn format_session_key(session_id: &str) -> String {
+        format!("session:{}", session_id)
+    }
 
-fn format_cache_key(url: &str) -> String {
-    format!("cache:{}", url)
+    fn format_cache_key(url: &str) -> String {
```
Comment (djc) [inline src/store.rs]: Please squash into original commit.

File: src/store.rs [hunk 7]
```diff
@@ -12,46 +12,46 @@ pub struct Store {
...
-    err.description().to_string()
-}
+    fn stringify_err(err: RedisError) -> String {
+        err.description().to_string()
+    }
 
-fn format_session_key(session_id: &str) -> String {
-    format!("session:{}", session_id)
-}
+    fn format_session_key(session_id: &str) -> String {
```
Comment (djc) [inline src/store.rs]: Please squash into original commit.

File: src/oidc.rs [hunk 8]
```diff
@@ -1,16 +1,12 @@
...
 use emailaddress::EmailAddress;
 use iron::Url;
 use serde_json::de::{from_reader, from_str};
 use serde_json::value::Value;
 use self::hyper::client::Client as HttpClient;
 use self::hyper::header::{
-    ContentType as HyContentType,
-    CacheControl as HyCacheControl,
-    CacheDirective as HyCacheDirective,
+    ContentType as HyContentType
```
Comment (djc) [inline src/oidc.rs]: This should go on a single line, without the `{}`.

File: src/oidc.rs [hunk 9]
```diff
@@ -47,8 +43,11 @@ pub fn request(app: &AppConfig, params: &QueryMap) -> Url {
     // Retrieve the provider's Discovery document and extract the
     // `authorization_endpoint` from it.
     let provider = &app.providers[&email_addr.domain];
-    let val = fetch_json_cached(app, &provider.discovery);
-    let config = val.as_object().unwrap();
+    let config_val: Value = {
```
Comment (djc) [inline src/oidc.rs]: These separate scopes are a good idea, but please put them in a separate commit from the one that introduces the caching methods, so that the changes for each commit are clearly described by the commit message.

File: README.rst [hunk 10]
```diff
@@ -45,7 +45,9 @@ Here's an example configuration file:
      ],
      "store": {
        "redis_url": "redis://127.0.0.1/5",
-       "expire_keys": 900
+       "expire_sessions": 900,
```
Comment (djc) [inline README.rst]: I'd prefer if each of these changes were in separate commits.

File: src/store.rs (hunk 1 above)
Comment (stephank) [inline src/store.rs]: I'm not sure how to do this. What I've tried is `impl From<RedisError> for String { ... }` and `impl Into<String> for RedisError { ... }`, but I run into:

    error[E0117]: only traits defined in the current crate can be implemented for arbitrary types

The error makes sense, but I'm not sure if there's some other path I should take.

Alternatively, we could define our own error type, and then provide conversions for that all the way to `IronError`? (Maybe something that should be part of #8?)

Comment (djc) [inline src/store.rs]: Yeah, I guess then implementing our own error type is probably the right solution. I'm thinking it would be an `enum` covering all the input error types (e.g. start with `RedisError`), then we can implement `From` for `RedisError` and `Into` for `IronError`.

**Comment by stephank** [comment]
@djc I reworked this branch (on top of current master) to try and address the design and history issues. Let me know what you think!

File: src/store.rs [hunk 11]
```diff
@@ -44,4 +46,9 @@ impl Store {
         }
         Ok(stored)
     }
+
+    fn format_session_key(session_id: &str)
```
Comment (djc) [inline src/store.rs]: This looks like it'll fit on one line?

File: src/email.rs [hunk 12]
```diff
@@ -71,28 +68,9 @@ pub fn request(app: &AppConfig, params: &QueryMap) -> Value {
         .build().unwrap();
     // TODO: Add support for authentication.
     let mut mailer = SmtpTransportBuilder::new(app.smtp.address.as_str()).unwrap().build();
-    let result = mailer.send(email);
+    mailer.send(email).unwrap();
     mailer.close();
 
-    // TODO: for debugging, this part returns a JSON response with some
```
Comment (djc) [inline src/email.rs]: These changes seem unrelated to the commit message, can you please put them in a separate commit?

File: src/lib.rs.in [hunk 13]
```diff
@@ -150,10 +151,9 @@ impl Handler for AuthHandler {
 
         } else {
 
-            // Email loop authentication. For now, returns a JSON response;
-            // empty if successful, otherwise contains an error.
-            let obj = email::request(&self.app, params);
-            json_response(&obj)
+            // Email loop authentication. For now, returns 204.
```
Comment (djc) [inline src/lib.rs.in]: This would also be better as a separate commit.

File: README.rst [hunk 14]
```diff
@@ -47,6 +47,9 @@ Here's an example configuration file:
        "redis_url": "redis://127.0.0.1/5",
        "expire_keys": 900
      },
+     "store_cache": {
```
Comment (djc) [inline README.rst]: I don't think the `StoreCache` abstraction should leak out into the configuration. Instead, we should have the different `expire_` config values under `store`, similar to what you had before.

File: src/store_cache.rs [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    pub expire_keys: usize, // Redis key minimum TTL, in seconds
+}
+
+impl StoreCache {
+    /// Fetch `url` from cache or using a HTTP GET request. The cache is stored in `store` with
+    /// `key`. The `session` is used to make the HTTP GET request, if necessary.
+    pub fn fetch_url(&self, store: &Store, key: CacheKey, session: &HttpClient, url: &str)
+                     -> BrokerResult<String> {
+
+        // Format the Redis key.
```
Comment (djc) [inline src/store_cache.rs]: IMO the comments in this function are a bit too fine-grained/are largely repeating the code (that is, they don't add that much value). I think they could be removed and/or changed to provide more rationale/why rather than what is happening.

**Comment by stephank** [comment]
@djc Rebased again, should address all feedback!

**Comment by djc** [comment]
Awesome, thank you! Especially for dealing with my critical reviews...

**Comment by stephank** [comment]
Woohoo, making great progress! 🎉

And no worries, I can be a stubborn fool when it comes to these things.
